File: integration/ts/covenant-vault.test.ts
/**
 * CovenantVault integration test — stateless contract with checkSig + checkPreimage.
 *
 * ## How It Works
 *
 * CovenantVault demonstrates a covenant pattern: it constrains HOW funds can be spent,
 * not just WHO can spend them. The contract checks:
 *   1. The owner's ECDSA signature (authentication via checkSig)
 *   2. The transaction preimage (via checkPreimage, which enables script-level
 *      inspection of the spending transaction)
 *   3. That the transaction outputs match the expected P2PKH script to the recipient
 *      with amount >= minAmount (enforced by comparing hash256(expectedOutput) against
 *      extractOutputHash(txPreimage))
 *
 * ### What is checkPreimage / OP_PUSH_TX?
 *   checkPreimage verifies a BIP-143 sighash preimage against the spending transaction.
 *   This is implemented via the OP_PUSH_TX technique: the unlocking script pushes
 *   both a preimage (the raw BIP-143 serialization) and an ECDSA signature computed
 *   with private key k=1 (whose public key is the generator point G). The locking
 *   script verifies this signature against the preimage, which proves the preimage
 *   is genuine. Once verified, the script can inspect transaction fields (outputs,
 *   amounts, etc.) by parsing the preimage — enabling covenant rules.
 *
 * ### Constructor
 *   - owner: PubKey — the ECDSA public key that must sign to spend
 *   - recipient: Addr — the hash160 of the authorized recipient's public key
 *   - minAmount: bigint — minimum satoshis that must be sent to the recipient
 *
 * ### Method: spend(sig: Sig, txPreimage: SigHashPreimage)
 *   The compiler inserts an implicit _opPushTxSig parameter before the declared params.
 *   The full unlocking script order is: <opPushTxSig> <sig> <txPreimage>
 *
 *   The contract constructs the expected P2PKH output on-chain using the recipient
 *   address and minAmount from its constructor parameters, then verifies it matches
 *   the actual transaction outputs via hash256(expectedOutput) == extractOutputHash.
 *
 * ### Spending Limitation
 *   Covenant spending requires constructing a transaction whose outputs exactly match
 *   what the contract expects (a P2PKH output to the recipient for minAmount satoshis).
 *   The SDK's generic call() creates default outputs that don't match. For real
 *   applications, developers would construct the spending transaction manually or use
 *   the SDK's raw transaction builder. The covenant logic is fully verified by the TS
 *   unit tests and conformance golden files.
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract } from 'runar-sdk';
import { createFundedWallet, createWallet } from './helpers/wallet.js';
import { createProvider } from './helpers/node.js';

describe('CovenantVault', () => {
  it('should compile the CovenantVault contract', () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('CovenantVault');
  });

  it('should deploy with owner, recipient, and minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      1000n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should deploy with zero minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      0n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy with large minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      100_000_000n, // 1 BTC in satoshis
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy with same key as owner and recipient', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const ownerAndRecipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      ownerAndRecipient.pubKeyHex,
      ownerAndRecipient.pubKeyHash,
      500n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
  });

  // NOTE: Covenant spending (spend() method) requires constructing a transaction
  // whose outputs exactly match the expected P2PKH output enforced by the contract:
  //   output = num2bin(minAmount, 8) || "1976a914" || recipient || "88ac"
  // The SDK's generic call() creates default outputs that don't satisfy this
  // covenant constraint. In production, developers use the SDK's raw transaction
  // builder to construct the exact required output. The covenant verification
  // logic is fully covered by the TS unit tests and conformance golden files.

  it('should reject spend with wrong signer (checkSig fails before covenant check)', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const recipient = createWallet();

    // Deploy with owner=walletA
    const { signer: ownerSigner, pubKeyHex: ownerPubKeyHex } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      ownerPubKeyHex,
      recipient.pubKeyHash,
      1000n,
    ]);

    await contract.deploy(provider, ownerSigner, {});

    // Call spend with walletB — checkSig will fail on-chain
    const { signer: wrongSigner } = await createFundedWallet(provider);

    await expect(
      contract.call('spend', [null, null], provider, wrongSigner),
    ).rejects.toThrow();
  });
});

File: integration/go/covenant_vault_test.go
//go:build integration

package integration

import (
	"encoding/hex"
	"testing"

	"runar-integration/helpers"

	"github.com/bsv-blockchain/go-sdk/script"

	runar "github.com/icellan/runar/packages/runar-go"
)

func deployCovenantVault(t *testing.T, owner, recipient *helpers.Wallet, minAmount int64) *runar.RunarContract {
	t.Helper()

	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/covenant-vault/CovenantVault.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	t.Logf("CovenantVault script: %d bytes", len(artifact.Script)/2)

	contract := runar.NewRunarContract(artifact, []interface{}{
		owner.PubKeyHex(),
		recipient.PubKeyHashHex(),
		int64(minAmount),
	})

	funder := helpers.NewWallet()
	helpers.RPCCall("importaddress", funder.Address, "", false)
	_, err = helpers.FundWallet(funder, 1.0)
	if err != nil {
		t.Fatalf("fund: %v", err)
	}

	provider := helpers.NewRPCProvider()
	signer, err := helpers.SDKSignerFromWallet(funder)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}

	_, _, err = contract.Deploy(provider, signer, runar.DeployOptions{Satoshis: 5000})
	if err != nil {
		t.Fatalf("deploy: %v", err)
	}

	return contract
}

func TestCovenantVault_Compile(t *testing.T) {
	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/covenant-vault/CovenantVault.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	if artifact.ContractName != "CovenantVault" {
		t.Fatalf("expected contract name CovenantVault, got %s", artifact.ContractName)
	}
	t.Logf("CovenantVault compiled: %d bytes", len(artifact.Script)/2)
}

func TestCovenantVault_Deploy(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 1000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=1000")
}

func TestCovenantVault_DeployZeroMinAmount(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 0)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=0")
}

func TestCovenantVault_DeployLargeMinAmount(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 100000000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=100000000 (1 BTC)")
}

func TestCovenantVault_DeploySameKey(t *testing.T) {
	wallet := helpers.NewWallet()
	contract := deployCovenantVault(t, wallet, wallet, 1000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with same key as owner and recipient")
}

func TestCovenantVault_ValidSpend(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	minAmount := int64(1000)

	contract := deployCovenantVault(t, owner, recipient, minAmount)

	// Get UTXO from SDK contract, convert for raw spending
	utxo := helpers.SDKUtxoToHelper(contract.GetCurrentUtxo())

	// Build spend TX with the EXACT output the covenant expects:
	// A P2PKH output to the recipient for minAmount satoshis.
	// The covenant verifies hash256(expectedOutput) == extractOutputHash(txPreimage),
	// where expectedOutput = num2bin(minAmount, 8) || "1976a914" || recipient || "88ac"
	recipientScript := recipient.P2PKHScript()
	spendTx, err := helpers.BuildSpendTx(utxo, recipientScript, minAmount)
	if err != nil {
		t.Fatalf("build spend: %v", err)
	}

	// Sign with owner's key
	sigHex, err := helpers.SignInput(spendTx, 0, owner.PrivKey)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	sigBytes, _ := hex.DecodeString(sigHex)

	// OP_PUSH_TX for checkPreimage
	opPushTxSigHex, preimageHex, err := helpers.SignOpPushTx(spendTx, 0)
	if err != nil {
		t.Fatalf("op_push_tx: %v", err)
	}
	opPushTxSigBytes, _ := hex.DecodeString(opPushTxSigHex)
	preimageBytes, _ := hex.DecodeString(preimageHex)

	// spend(sig: Sig, txPreimage: SigHashPreimage)
	// Compiler inserts implicit _opPushTxSig before declared params.
	// Unlocking script order: <opPushTxSig> <sig> <txPreimage>
	unlockHex := helpers.EncodePushBytes(opPushTxSigBytes) +
		helpers.EncodePushBytes(sigBytes) +
		helpers.EncodePushBytes(preimageBytes)

	unlockScript, _ := script.NewFromHex(unlockHex)
	spendTx.Inputs[0].UnlockingScript = unlockScript

	txid := helpers.AssertTxAccepted(t, spendTx.Hex())
	helpers.AssertTxInBlock(t, txid)
}

// TestCovenantVault_WrongOutput_Rejected verifies the covenant rejects a transaction
// whose output doesn't match the expected P2PKH to the registered recipient.
func TestCovenantVault_WrongOutput_Rejected(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	minAmount := int64(1000)

	contract := deployCovenantVault(t, owner, recipient, minAmount)

	utxo := helpers.SDKUtxoToHelper(contract.GetCurrentUtxo())

	// Build TX with wrong output: send to OWNER instead of RECIPIENT.
	// The covenant expects output to recipient — this should fail the
	// hash256(expectedOutput) == extractOutputHash(txPreimage) check.
	wrongScript := owner.P2PKHScript()
	spendTx, err := helpers.BuildSpendTx(utxo, wrongScript, minAmount)
	if err != nil {
		t.Fatalf("build spend: %v", err)
	}

	sigHex, err := helpers.SignInput(spendTx, 0, owner.PrivKey)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	sigBytes, _ := hex.DecodeString(sigHex)

	opPushTxSigHex, preimageHex, err := helpers.SignOpPushTx(spendTx, 0)
	if err != nil {
		t.Fatalf("op_push_tx: %v", err)
	}
	opPushTxSigBytes, _ := hex.DecodeString(opPushTxSigHex)
	preimageBytes, _ := hex.DecodeString(preimageHex)

	// Unlocking script order: <opPushTxSig> <sig> <txPreimage>
	unlockHex := helpers.EncodePushBytes(opPushTxSigBytes) +
		helpers.EncodePushBytes(sigBytes) +
		helpers.EncodePushBytes(preimageBytes)

	unlockScript, _ := script.NewFromHex(unlockHex)
	spendTx.Inputs[0].UnlockingScript = unlockScript

	helpers.AssertTxRejected(t, spendTx.Hex())
}
